### ISSUE TITLE
FreeBSD contigmem driver crashes when memory cannot be allocated

### ISSUE THREAD

**Issue by pawelkax** [spdk/spdk#1262]
To reproduce this error:
1. Create virtual machine with freebsd 12.1
2. Compile spdk sources
3. Run setup.sh

Output in console:
```
[vagrant@bazinga ~/spdk_repo/spdk]$ sudo ./scripts/setup.sh
hw.nic_uio.bdfs="0:7:0"
hw.contigmem.num_buffers="8"
hw.contigmem.buffer_size="268435456"
```
Output in vnc:
![image](https://user-images.githubusercontent.com/28859087/76021623-d62e2b00-5f25-11ea-930c-5fbd0a34e0a0.png)

**Comment by pawelkax** [comment]
I would like to add that after reboot the machine setup.sh is working.

**Comment by darsto** [comment]
A quick git log over DPDK's contigmem shows only 1 change since 2019 and that's a FreeBSD 13 compilation fix - unrelated here.

We should report this to DPDK if we ever see it again.

https://github.com/DPDK/dpdk/commit/03bff90ccfd6ad347f1e50901caf47890872ca28

**Comment by jimharris** [comment]
@pawelkax - are you able to intermittently reproduce this?

**Comment by pawelkax** [comment]
@jimharris It happens every time when I'm compile spdk for the first time ad use setup.sh.

**Comment by jimharris** [comment]
Do you see any other errors before this crash?

I looked at module_register_init() to understand why we are in the sys_kldload path but crashing in contigmem_unload.  If the MOD_LOAD fails, module_register_init() will immediately call MOD_UNLOAD.  So it seems to indicate that the load failed, and I'm wondering if there are any error messages which would indicate why.

But I do see the bug in the contigmem driver - if MOD_LOAD fails, we free any buffers that were able to be allocated.  But we don't set the buffer pointers back to NULL.  So then when the MOD_UNLOAD happens, we try to free those non-NULL buffers again, causing this crash.  Can you try the attached patch?

[patch.txt](https://github.com/spdk/spdk/files/4299261/patch.txt)

**Comment by pawelkax** [comment]
I tried to run setup.sh with patch.txt.
In console I got:
 11:10:52       # ./scripts/setup.sh
kldunload: can't find file nic_uio.ko
hw.nic_uio.bdfs="0:7:0"
hw.contigmem.num_buffers="8"
hw.contigmem.buffer_size="268435456"
kldload: an error occurred while loading module contigmem.ko. Please check dmesg(8) for more details.

 In dmesg I got error:
nic_uio0: <DPDK PCI Device> mem 0xfeb92000-0xfeb93fff,0xfeb98000-0xfeb98fff irq 11 at device 7.0 on pci0
 0: virt=0xfffffe002ce00000 phys=0x110000000
 1: virt=0xfffffe003d200000 phys=0x120000000
 2: virt=0xfffffe004d600000 phys=0x90000000
 3: virt=0xfffffe005da00000 phys=0xa0000000
contigmalloc failed for buffer 4
module_register_init: MOD_LOAD (contigmem, 0xffffffff82897000, 0) error 12
nic_uio0: detached

**Comment by jimharris** [comment]
How much memory do you have in this VM?

**Comment by pawelkax** [comment]
I had 4GB. Increased to 12GB and it works now.

**Comment by jimharris** [comment]
OK.  The contigmem driver load can fail if it cannot find enough contiguous memory - which may happen if you do something like compile SPDK before loading contigmem.  That explains why it works when you either add more memory, or reboot the VM and retry without first compiling SPDK.

But the crash still needs to be fixed.  I'm going to change the title of this bug to reflect the crash, and we'll keep this open until a version of my patch gets checked into our DPDK submodule.  I'll also work on getting that upstreamed.

**Comment by jimharris** [comment]
https://review.spdk.io/gerrit/c/spdk/dpdk/+/1200

**Comment by pawelkax** [comment]
@jimharris  I can confirm that spdk tests on freebsd 12 are working without your patch.txt when I have enough memory.
But after compiling spdk sources I cannot do scripts/setup.sh:
kldunload: can't find file nic_uio.ko
hw.nic_uio.bdfs="0:6:0"
hw.contigmem.num_buffers="8"
hw.contigmem.buffer_size="268435456"
kldload: an error occurred while loading module contigmem.ko. Please check dmesg(8) for more details.
In dmesg:
smbus0: <System Management Bus> on intsmb0
nic_uio: hw.nic_uio.bdfs = '0:6:0'
nic_uio_load: detaching and storing dev=0xfffff8000386f900
nvd0: detached
nvme0: detached
nic_uio0: <DPDK PCI Device> mem 0xfeb92000-0xfeb93fff,0xfeb97000-0xfeb97fff irq 10 at device 6.0 on pci0
contigmalloc failed for buffer 0
module_register_init: MOD_LOAD (contigmem, 0xffffffff8281f000, 0) error 12

After reboot setup.sh is working.

**Comment by jimharris** [comment]
HI @pawelkax - I think that's fine that setup.sh doesn't work after you compile SPDK sources.  Compiling the sources seems to fragment memory enough that we can't allocate even 1 256MB buffer.  It just means that on FreeBSD, we should always load contigmem (or run setup.sh) immediately after booting.

**Comment by jimharris** [comment]
Pushed to dev@dpdk.org also:

https://patches.dpdk.org/patch/66470/

**Comment by pawelkax** [comment]
@jimharris Ok, thank you for explaining.

**Comment by jimharris** [comment]
DPDK patch was merged to our spdk-19.11 branch.

https://review.spdk.io/gerrit/c/spdk/spdk/+/1217 moves the submodule to include this commit.

**Comment by darsto** [comment]
Fix merged
